### DISCUSSION TITLE
Filtering: How to filter rows that have equal values in 2 columns

### DISCUSSION THREAD

**Discussion by tglynx** [Budibase/budibase#11178] [Help] [answered]
Hi,

I have a table with lots of rows that contains the columns 'InsuranceNumber' and 'LicenseNumber' (among many others as an example). I want to help the maintainers of the dataset to quickly identify 'wrong' rows by providing specific views on common errors in the data.

As an example i want to filter rows where InsuranceNumber equals LicenseNumber in a dataprovider. (this is a common error as the License Number gets often mixed up with the Insurance Number in this case here). 
However, i am not sure if this is possible or how to do it. 

<img width="1021" alt="Bildschirm­foto 2023-07-10 um 13 49 35" src="https://github.com/Budibase/budibase/assets/30650435/bd35424f-704b-4f0c-8759-5fe7e8ab4fbc">

What would be the correct binding here to check for equal 'InsuranceNumber' in the same row? 

Thanks, 
Michael

**Comment by ConorWebb96** [comment]
Hey @tglynx,

What type of database are you working with here? 

From what I understand you may be best using a query based data source and then using a custom query to search through the entire database and pull out relevant rows matching your parameters. If you are using the internal database I'd highly recommend switching to a query based one as the internal database isn't really suited for this type of filtering.

**MySQL Example query that could do this**
```
SELECT *
FROM your_table_name
WHERE InsuranceNumber = LicenseNumber;
```

From here you could then use this query as the source for your data provider. Allowing you to iterate threw all rows that are relevant.

I hope this makes sense.

**Reply by tglynx** [reply]
Hi Conor, 

sorry, for me not mentioning, i am using the bb internal db. Obviously SQL does not work there.
Unfortunately i started this project in the bb internal db not aware of the limitations i may hit at that time.

I have thought about switching to MySQL already but i fear that it may be lots of work...

**Comment by R2bEEaton** [comment, marked as answer]
I'd highly recommend using a Custom Query to do this, but if you're on the internal database and have a small number of rows, here's a workaround:

Create a structure like this where Outer and Inner point to the same table.

![image](https://github.com/Budibase/budibase/assets/34921506/7849ff39-d2f3-4ecb-91bc-ca401bae922b)

Make sure the Outer provider's limit is as large as the number of rows in your table. Then on the Inner data provider, add this filter:

![image](https://github.com/Budibase/budibase/assets/34921506/c142f938-4193-4120-9049-e379fd7ce24b)

Where the JavaScript binding is:

```javascript
var rows = $("Outer.Rows");
var output = [];

for (let i = 0; i < rows.length; i++) {
  if (rows[i].InsuranceNumber== rows[i].LicenseNumber) output.push(rows[i]._id);
}

return output;
```

This has the effect of displaying the rows from the table which have the same InsuranceNumber and LicenseNumber in the Outer data provider. I hope this helps as a last-ditch effort, but I'd still highly recommend switching to a query-based data source like the other answer suggests, which would be much more performant than this option.

**Reply by tglynx** [reply]
Hi R2bEEaton,

nice approach, i will give it a try! With this logic i may also be able to solve some of my other requirements. 

thanks,
Michael

**Reply by tglynx** [reply]
Just wanted to let you know! Works as advertised 🥇

**Reply by R2bEEaton** [reply]
Wonderful! I'm super glad it's working!

Just a word of warning in case things break later down the line:
- Large row counts will slow this down
- If the total number of rows in the table exceeds the limit in the Outer provider, your filter may not retrieve all the rows you want

Without a doubt, @ConorWebb96's answer is the best and correct way to do it, and switching to some external query database may help you get around similar limitations in the future that you haven't considered yet. I had a similar experience - I kept hacking together solutions but had to end up switching to MSSQL. But if the conditions I mentioned are fine for you, then go right ahead 😃.

The good news is that if you decide to move over to MySQL or some other RDBMS, it's not too bad because Budibase was very smart about keeping `_id` rows around. It essentially detects your primary keys (including composite keys) and makes them into a special ID you can pass around in URLs and stuff like normal using the `_id` field. And depending on your DB and site structure a migration may result in impressive performance gains: #8331.

Enjoy!
